File: Java/DecodeWays.java
class Solution {
    public int numDecodings(String s) {
        int dp[] = new int[s.length()];
        if(s.charAt(0)!='0')
        dp[0] = 1;
        
        for(int i=1;i<dp.length;i++)
        {
            char a = s.charAt(i);
            char b = s.charAt(i-1);
            
            if(a=='0'&&b=='0')
            {
                dp[i] = 0;
            }
            else if(a=='0'&&b!='0')
            {
                if(b=='1'||b=='2')
                {
                    dp[i] = (i>=2 ? dp[i-2]:1);
                }
            }
            else if(a!='0'&&b=='0')
            {
                dp[i] = dp[i-1];
            }
            else
            {
                int x = Integer.parseInt(s.substring(i-1,i+1));
                if(x<=26)
                {
                    dp[i] = dp[i-1] + (i>=2?dp[i-2]:1);
                }
                else
                {
                    dp[i] = dp[i-1];
                }
            }
        }
        return dp[s.length()-1];
    }
}

File: CPP/Decode ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        int n=s.size();
        int dp[n][2];
        
        if(s[0]=='0')
            return 0; 
        
        dp[0][0]=1;dp[0][1]=0;
        
        for(int i=1;i<n;i++){
            if(s[i]=='0'){
                if(s[i-1]=='1' || s[i-1]=='2'){
                    dp[i][0] = 0 ;
                    dp[i][1] = dp[i-1][0] ;
                }
                
                else 
                    return 0;
            }
            
            else{
                dp[i][0] = dp[i-1][0] + dp[i-1][1] ;
                dp[i][1] = 0;
                if(stoi(s.substr(i-1,2))<=26)
                    dp[i][1] = dp[i-1][0] ;
            }
             
        }
        return dp[n-1][0]+dp[n-1][1];
    }
};